### DISCUSSION TITLE
BWA-MEM alt-aware mode

### DISCUSSION THREAD

**Discussion by rhughwhite** [uclahs-cds/pipeline-align-DNA#262] [Ideas]
## Alt-aware alignment
In addition to the primary contigs/chromosmes, reference genomes contain ALT contigs which represent common regions of complex variation.

BWA-MEM has an 'alt-aware' mode which causes it to prioritize aligning reads to primary contigs, with ALT contig alignments being restricted to reads that cannot be mapped to primary contigs. Additionally in alt-aware mode, quality scores are calculated across primary and matching contigs. Alt-aware mode is likely preferable in most cases since many downstream tools simply ignore alt-mapping reads.

The alt-aware mode is active by default and can be turned off with the `-J` parameter. However alt-aware mode is only used if a .alt index file is included in the same directory with the same prefix as the reference fasta. Unlike other bwa index files the .alt file isn't user-generated but is instead downloaded as part of a [bwa-kit package](https://sourceforge.net/projects/bio-bwa/files/bwakit/).

## Problem

Previously in the lab we had a .alt file included with the BWA-MEM index files and alignments were being performed in alt-aware mode. During the switch from BWA-MEM to BWA-MEM2 it seems like the .alt file was lost and so subsequent alignments have not been alt-aware. The main effect is that these alignments will have increased alt contig mapping reads, though the extent may vary based on sample ancestry.

## Solution
It would be good to add the .alt file to `/hot/ref/tool-specific-input/BWA-MEM2-2.1.0/GRCh38-BI-20160721/` and include the `-J` parameter as an option (possibly with a warning if it is used). An additional option would be to provide an index set with alt contig masking which would fully inhibit mapping to alt contigs (this might be lower priority) - https://support-docs.illumina.com/SW/DRAGEN_v310/Content/SW/DRAGEN/RefGenHashALT.htm

See this GATK resource for more:
https://gatk.broadinstitute.org/hc/en-us/articles/360037498992--How-to-Map-reads-to-a-reference-with-alternate-contigs-like-GRCH38

**Comment by tnyamaguchi** [comment]
Thanks for reporting this @rhughwhite . It looks like the `-j` option is available in BWA-MEM2 as well.
`-j            treat ALT contigs as part of the primary assembly (i.e. ignore <idxbase>.alt file)`
Would you be able to test this out and see if it works as expected with BWA-MEM2 if you haven't?

**Comment by lydiayliu** [comment]
As per request from @graceooh, I will post the process of how I found that my samples were not run in `alt-aware` mode.

I have head and neck cancer samples that have been run with `bwa mem 0.7.17` and `align-DNA 3.0.0` that subsequently received top-up sequencing. @rhughwhite realigned each sample with the original FASTQs together with the top-up FASTQs using `bwamem 2.2.1` and `align-DNA 9.0.0`. We then proceeded to run Battenberg, and samples were failing due to a previously reported error (https://github.com/Wedge-lab/battenberg/issues/37), which I diagnosed as due to stretches of primary chromosomes having read depth of zero.

Since we realigned with original as well as top-up FASTQs, read count could only theoretically increase. Yet in the original BAMs aligned with `bwa mem 0.7.17` and `align-DNA 3.0.0` we had read depth in the problematic regions.

@rhughwhite then proceeded to eliminate every possible variable but we could not reproduce the previous results. We reasoned that the difference could only be due to the reference genome and the index, since the reference files have been reorganized.

@maotian06 pointed us in the right direction by linking this article: https://gatk.broadinstitute.org/hc/en-us/articles/360037498992--How-to-Map-reads-to-a-reference-with-alternate-contigs-like-GRCH38. It made us realize that the original alignment was in `alt-aware` mode which required the special `.alt` index file. The `.alt` index file cannot be generated by `bwa index` and has to be downloaded as part of `bwakit`. After adding the `.alt` index file to our `bwa mem 0.7.17` index directory, @rhughwhite was able to reproduce the prior `alt-aware` alignments.

We suspect that the issue was especially prominent for my samples because they are of East Asian ancestry. For those who had no trouble performining downstream analyses, the results are unlikely to differ significantly in `alt-aware` mode as the alt contigs only account for a small percentage of bases of the primary chromosomes. Nevertheless, it does help read depth in those regions tremendously as multiply mapped reads are given low or zero MAPQ scores.

**Comment by tnyamaguchi** [comment]
@rhughwhite @lydiayliu Thank you guys for providing the information. We've looked into this and there are pros and cons to enable the alt-aware feature but given that we're processing samples with a diverse genetic background, I think in most cases, enabling alt-aware mode would be beneficial. We don't want to miss somatic variant calls due to false low mapping quality(MQ) regions, as many somatic variant calling algorithms use MQ filtering. 

However, a potential drawback is that BWA-MEM2 alt-aware mode might not work as expected compared to the BWA-MEM alt-aware mode as shown here https://gatk.broadinstitute.org/hc/en-us/articles/360037498992--How-to-Map-reads-to-a-reference-with-alternate-contigs-like-GRCH38#1.
Also, this benchmark highlighted some discrepancies although the article is somewhat outdated - https://blog.dnanexus.com/2020-03-10-bwa-mem2-review

To support the alt-aware mode on the cluster, we can go ahead and create a new directory here - `/hot/ref/tool-specific-input/BWA-MEM2-2.2.1/GRCh38-BI-20160721/alt-aware` for now because BWA-MEM2 defaults to using the `.alt` file if available. We can keep the original index directory as is to avoid unintentionally mixing both modes. @yashpatel6 can you please help with this as you've already created the indices?

@rhughwhite @lydiayliu @graceooh @maotian06 Feel free to share your benchmarking results in this thread if any. 

Let me know if you guys have any thoughts or concerns.

**Reply by yashpatel6** [reply]
> To support the alt-aware mode on the cluster, we can go ahead and create a new directory here - `/hot/ref/tool-specific-input/BWA-MEM2-2.2.1/GRCh38-BI-20160721/alt-aware` for now because BWA-MEM2 defaults to using the `.alt` file if available. We can keep the original index directory as is to avoid unintentionally mixing both modes. @yashpatel6 can you please help with this as you've already created the indices?

Done, indices with the `.alt` file should now be available at `/hot/ref/tool-specific-input/BWA-MEM2-2.2.1/GRCh38-BI-20160721/alt-aware`!
